subsys/net/ip/tcp.c

```diff
@@ -1126,6 +1126,8 @@ static void tcp_resend_data(struct k_work *work)
 	bool conn_unref = false;
 	int ret;
 
+	k_mutex_lock(&conn->context->lock, K_FOREVER);
```

Comment:
Thanks for investigating the issue!

As far as I understand, locking `net_context` here should be no harm (unless I miss something), but I'm wondering if it's actually desired here. It indeed would help to give the TCP thread a chance to run in certain conditions, but is not robust enough to work in all cases.

The networking stack in general can be configured to run its threads in either cooperative or preemptive mode. In case it's configured into cooperative mode, the TCP thread gets the highest priority possible in the system, so the scheduler will indeed give it a chance to run if it's waiting on the `net_context` mutex and when it gets freed by the TX thread. In case the stack is working in a preemptive mode however, the TX thread can still get higher priority than the TCP workqueue, so even if it's pending on the `net_context` mutex, the scheduler will not give it a chance to run if the TX thread with non-blocking socket is spinning trying to resend.

With this solution, the TX thread will still keep spinnging inbetween resend calls, hogging the CPU, is that correct?

I think we should rather try to resolve the core issue causing the TCP thread starvation, i. e. the transmitting thread hogging the CPU when window is full. This means implementing POLLOUT for TCP socket properly, so that non-blocking socket can use it to monitor whether the window is full or not, effectively preventing it from busy looping.

Reply:
@rlubos  You are totally right, I didn't consider the preemptive case. 
One adjustment (which I probably will have to use in our fork of Zephyr for production) would be to check priorities and then to either lock context mutex or not (or just check CONFIG_NET_TC_THREAD_COOPERATIVE)

I see two core issues here.

One is indeed POLLOUT being not implemented. 
However, then we rely on user calling select()/poll(), which most people would forget at first.
Heck, even http_client implementation here in Zephyr doesn't do that. 

Another issue is the order of allocating and unreferencing network buffers. 
I would expect that resend doesn't need to do any allocation. Then the problem would disappear on itself.
As far as I understood, it doesn't need extra memory conceptually, but it does extra allocation, temporarily consuming extra memory (allocate, copy, add to send_queue, then unreference old packets and buffers).

Reply:
> One is indeed POLLOUT being not implemented.
However, then we rely on user calling select()/poll(), which most people would forget at first.
Heck, even http_client implementation here in Zephyr doesn't do that.

Actually, I think it's sane to assume that if the user is using a non-blocking socket, it would use poll()/select() to synchronize, otherwise he asks for a trouble (same as with non-blocking recv()). And indeed, `http_client` from Zephyr should take that into consideration, since the user can configure a socket to be a non-blocking (this needs fixing).

> I would expect that resend doesn't need to do any allocation.

The low-level networking in Zephyr works in a way, that the network driver frees the provided net packet after sending, therefore TCP needs to create a copy of a packet before transmitting (in case it expects that the data packet may need retransmission). Changing this behaviour would require to modify the whole networking stack, which is not feasible IMO.

Reply:
@rlubos 

>With this solution, the TX thread will still keep spinnging inbetween resend calls, hogging the CPU, is that correct?

Actually, on second thought, it should work for preemptive as well (will test it ASAP). 

Because of the lock. If we keep flooding the socket with requests, once the lock is free, resend will be called if it was due.
And then another context_sendto will be called only once resend is finished, regardless of the priority, because context_sendto is surrounded with context locks.

Reply:
@rlubos 

Even in case of preemptive solution, network takes highest preemptive priority, 0:

```
#if IS_ENABLED(CONFIG_NET_TC_THREAD_COOPERATIVE)
#define THREAD_PRIORITY K_PRIO_COOP(0)
#else
#define THREAD_PRIORITY K_PRIO_PREEMPT(0)
#endif
```
So in some cases it will work for pre-emptive as well.

To generalize, my proposal will work in the following cases (pseudocode):
```
(CONFIG_NET_TC_THREAD_COOPERATIVE) && (NET_CONTEXT_SEND_PRIO >= 0)
||
(CONFIG_NET_TC_THREAD_PREEMPTIVE) && (NET_CONTEXT_SEND_PRIO > 0)
```

Reply:
Yeah, but the application is free to configure its thread as cooperative, and it has nothing to do with `CONFIG_NET_TC_THREAD_*` options. So we should also consider a scenario where networking threads are preemptive, but the application is cooperative, and that's what I was concerned about initially.

Reply:
Agreed.
But you can confirm, that in cases I mention it will work, right? Because we'd use in our fork until proper polling is implemented. By the way thanks a lot for starting implementation!

I also agree that implementing (and using) poll will prevent the issue from happening.

My only concern is how to communicate to the futures users that Data buffer allocation error is caused by polling not working/not used, because I spent quite some time before understanding what's really going on. 

May be we could add this case to Q&A as well, and definitely we should add call to poll in http_client.

Reply:
> But you can confirm, that in cases I mention it will work, right? Because we'd use in our fork until proper polling is implemented. By the way thanks a lot for starting implementation!

Yes, in those certain cases it should help.

Reply:
Thanks for taking time to look into this as well!

I guess then once polling is implemented we can close this PR as unnecessary.